src/core/lib/gprpp/ref_counted.h

```diff
@@ -58,6 +60,73 @@ class NonPolymorphicRefCount {
   ~NonPolymorphicRefCount() {}
 };
 
+// RefCount is a simple atomic ref-count.
+//
+// This is a C++ implementation of gpr_refcount, with inline functions. Due to
+// inline functions, this class is significantly more efficient than
+// gpr_refcount and should be preferred over gpr_refcount whenever possible.
+//
+// TODO(soheil): Remove gpr_refcount after submitting the GRFC and the paragraph
+//               above.
+class RefCount {
+ public:
+  using Value = intptr_t;
+
+  // `init` is the initial refcount stored in this object.
+  explicit RefCount(Value init = 1) : val_(init) {}
+
+  // Reinitializes the ref-count.
+  void Reset(Value init) { val_.store(init, std::memory_order_relaxed); }
+
+  // Increases the ref-count by `n`.
+  void Ref(Value n = 1) { val_.fetch_add(n, std::memory_order_relaxed); }
+
+  // Increases the ref-count by `n` with a trace log.
+  void Ref(const DebugLocation& location, const char* reason,
```

Comment:
This seems like a fairly ugly method for the public API.  If this is only needed by `RefCountedWithTracing`, then I suggest just making that a friend of this class and having it directly access `val_`.

Reply:
I ack that it's not a pleasant looking API but, unfortunately, this is a very common pattern in gRPC and some can not be easily transferred to RefCountedWithTracing.

Here is an example, but you can find them by "gpr_log.*ref":
https://github.com/grpc/grpc/blob/b41a44dc2e001358602e8e885c674ddc49387edd/src/core/ext/transport/chttp2/transport/chttp2_transport.cc#L205

I'm happy to add the get() method back and add the friend class, but will have to do that for quite a few occurrences. Please just let me know.

Reply:
Is this ever used outside `#ifndef NDEBUG` ? If so, perhaps that's the tack to account for the ugliness? To leave it in but only as a clear debugging option?

Reply:
Okay, I guess let's go back to `get()`, and don't worry about making it private.  If there are a lot of callers, public makes sense, and `get()` is a much cleaner thing to have in the public API than this is.